stat_fem/ForcingCovariance.py

```diff
@@ -1,2 +1,276 @@
+        """
+
+        if self.is_assembled:
+            return
+
+        G_dict, nnz = self._compute_G_vals()
+
+        self.G = PETSc.Mat().create(comm=self.comm)
+        self.G.setType('aij')
+        self.G.setSizes(((self.nx_local, -1), (self.nx_local, -1)))
+        self.G.setPreallocationNNZ(nnz)
+        self.G.setFromOptions()
+        self.G.setUp()
+
+        for key, val in G_dict.items():
+            self.G.setValues(np.array(key, dtype=PETSc.IntType), val[1], val[0])
+
+        self.G.assemble()
+
+        self.is_assembled = True
+
+    def destroy(self):
+        """
+        Destroy allocated covariance forcing matrix
+
+        Deallocates memory for an assembled forcing matrix. If memory has not been allocated, returns
+        with no effect. No inputs, no return value.
+        """
+
+        if not self.is_assembled:
+            return
+
+        self.G.destroy()
+
+    def mult(self, x, y):
+        """
+        Perform matrix multiplication with firedrake vector
+
+        The principal operation done with a ``ForcingCovariance`` object is matrix multiplication
+        with a vector. This method is a wrapper to the underlying method for the PETSc object, which
+        modifies the ``y`` vector argument in-place. The provided vectors must be Firedrake ``Vector``
+        objects.
+
+        If the underlying covariance sparse matrix has not been assembled, it will be assembled prior
+        to performing the matrix multiplication.
+
+        :param x: Firedrake vector that will be multiplied with the covariance matrix. Must be a distributed
+                  vector with the same row allocations as the covariance matrix.
+        :type x: Vector
+        :param y: Firedrake vector that will hold the result of multiplying the input vector with the
+                  covariance matrix. Must be a distributed vector with the same row allocations as the
+                  covariance matrix. Modified in place.
+        :type y: Vector
+        :returns: None
+        """
+
+        assert isinstance(x, Vector), "x must be a firedrake vector"
+        assert isinstance(y, Vector), "y must be a firedrake vector"
+
+        if not self.is_assembled:
+            self.assemble()
+
+        with x.dat.vec_ro as xtmp:
+            with y.dat.vec as ytmp:
+                self.G.mult(xtmp, ytmp)
+
+    def get_nx(self):
+        """
+        Return number of global nodal DOFs for FEM
+
+        Returns the number of global nodal DOFs for the FEM over all processes.
+
+        :returns: number of global DOFs
+        :rtype: int
+        """
+
+        return self.nx
+
+    def get_nx_local(self):
+        """
+        Return number of local nodal DOFs for FEM
+
+        Returns the number of local nodal DOFs for the FEM for the current process.
+
+        :returns: number of local DOFs
+        :rtype: int
+        """
+
+        return self.nx_local
+
+    def __str__(self):
```

Comment:
You probably already know this, but `__repr__` is often used as a string method because `__repr__` is designed to convert a (possibly mutable) object into something hashable (must be immutable) so the object can be used as the key in a dictionary. Strings are commonly used for this as they are immutable, so long as they can be used to identify an object with some hope of uniqueness. My worry in defining one here is that the way I defined the string is hardly likely to be unique, so it's probably not well suited for hashing.